### PR TITLE
(test) O3-3937 Add test coverage around form readonly mode

### PR DESCRIPTION
**Pull request by CynthiaKamau**
## Requirements

- [ ] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary

O3-3937 Add test coverage around form readonly mode

## Related Issue

https://openmrs.atlassian.net/browse/O3-3937

### CODE REVIEW

File: src/form-engine.test.tsx [hunk 1]
```diff
@@ -215,6 +216,33 @@ describe('Form engine component', () => {
...
+
+      expect(visitPunctualityTextbox).toBeInTheDocument();
+      expect(visitPunctualityTextbox).toHaveClass('cds--text-input');
+
+      const visitPunctualityWrapper = visitPunctualityTextbox.closest('.cds--text-input-wrapper');
+      expect(visitPunctualityWrapper).toHaveClass(
+        'cds--form-item cds--text-input-wrapper cds--text-input-wrapper--readonly',
+      );
+    });
+  });
```
Comment (samuelmale) [inline src/form-engine.test.tsx:244]: I love tests but this test is supposed to be testing a form rendered in "view" mode. I've updated the ticket to clearly reflect this. (sorry for the confusion)

> The form-engine supports three major modes: enter (new), edit and view (readonly). This ticket is meant for adding test coverage of the form’s behaviour while in view mode.

Comment (CynthiaKamau) [reply src/form-engine.test.tsx:244]: No worries, let me do a rework

File: src/form-engine.test.tsx [hunk 2]
```diff
@@ -842,15 +844,68 @@ describe('Form engine component', () => {
...
+
+      const blankFields = screen.getAllByText(/\(Blank\)/i);
+      blankFields.forEach((blankField) => {
+        expect(blankField).toBeInTheDocument();
+      });
+
+      const inputs = screen.queryAllByRole('textbox');
+      inputs.forEach((input) => {
+        expect(input).toHaveAttribute('readonly');
+      });
```
Comment (samuelmale) [inline src/form-engine.test.tsx:893]: Can you also assert that the "Save" button is disabled?

File: src/utils/forms-loader.test.ts [hunk 3]
```diff
@@ -127,6 +127,15 @@ describe('Forms loader - getForm', () => {
...
                 },
+                {
+                  id: 'sampleQuestion',
+                  label: 'Question title',
+                  questionOptions: {
+                    concept: '193e8a03-1581-46e9-b0a4-17c55fa2649f',
+                    rendering: 'text',
+                  },
+                  type: 'obs',
+                },
```
Comment (samuelmale) [inline src/utils/forms-loader.test.ts:138]: Do you need this?

File: src/utils/forms-loader.test.ts [hunk 4]
```diff
@@ -246,6 +255,15 @@ describe('Forms loader - getForm', () => {
...
```
Comment (samuelmale) [inline src/utils/forms-loader.test.ts:266]: Ditto

File: src/form-engine.test.tsx [hunk 5]
```diff
@@ -842,15 +844,68 @@ describe('Form engine component', () => {
...
+      const visitTypeDropdown = screen.getByRole('combobox', {
+        name: /visit type/i,
+      });
+      const visitPunctualityTextbox = screen.getByLabelText(/visit punctuality/i);
+
+      expect(visitTypeDropdown).toBeInTheDocument();
+      expect(visitTypeDropdown).toHaveClass('cds--list-box__field');
+
+      const visitTypeWrapper = visitTypeDropdown.closest('.cds--dropdown');
+      expect(visitTypeWrapper).toHaveClass('cds--dropdown cds--dropdown--readonly cds--list-box');
```
Comment (samuelmale) [inline src/form-engine.test.tsx:862]: What happens if the CSS names change? Can you use attributes instead?

**Review by samuelmale — CHANGES_REQUESTED** [review]

File: src/form-engine.test.tsx (hunk 2 above)
Comment (CynthiaKamau) [reply src/form-engine.test.tsx:880]: Added it on line 884

File: src/utils/forms-loader.test.ts (hunk 4 above)
Comment (CynthiaKamau) [reply src/utils/forms-loader.test.ts]: Screenshot 2024-10-18 at 17 38 57


Removing this block brings this error

**Pull request merged**